### PR TITLE
[Prediction] Modernize polypred a little by introducing ctor.

### PR DESCRIPTION
**Pull request by kumpera**
In order to introduce a ctor, and pave the way for a dtor, we need to make the
type copy/assignment safe.

Since v_array is not copy safe, so is polyprediction. So I decided to make it
not copyable/assignable, only moveable.

This change introduce a new guard, stash_guard, that borrows the name from the same name
git command. It's used all over the place to replace code that manually save/restore example::pred.

With this change in, it should be possible to move cleanup from ~example() to ~polyprediction.

### CODE REVIEW

File: vowpalwabbit/nn.cc [hunk 1]
```diff
@@ -394,7 +394,8 @@ void multipredict(nn& n, single_learner& base, example& ec, size_t count, size_t
     else
       predict_or_learn_multi<false, false>(n, base, ec);
     if (finalize_predictions)
-      pred[c] = ec.pred;
+      pred[c] = std::move(ec.pred);  // TODO: this breaks for complex labels because = doesn't do deep copy! (XXX we
+                                     // "fix" this by moving)
```
Comment (olgavrou) [inline vowpalwabbit/nn.cc:398]: nit: can we remove the TODO now?

File: vowpalwabbit/learner.h [hunk 2]
```diff
@@ -243,7 +243,8 @@ struct learner
       {
         learn_fd.predict_f(learn_fd.data, *learn_fd.base, (void*)&ec);
         if (finalize_predictions)
-          pred[c] = ec.pred;  // TODO: this breaks for complex labels because = doesn't do deep copy!
+          pred[c] = std::move(ec.pred);  // TODO: this breaks for complex labels because = doesn't do deep copy! (XXX we
```
Comment (olgavrou) [inline vowpalwabbit/learner.h:246]: nit: same here?

File: vowpalwabbit/example.h [hunk 3]
```diff
@@ -50,6 +50,15 @@ inline void delete_scalars(void* v)
...
 {
+  polyprediction();
+
+  // disable those two since v_array is not cleanly copy-able
+  polyprediction(polyprediction&&) = default;
+  polyprediction& operator=(polyprediction&&) = default;
+
+  polyprediction(const polyprediction&) = delete;
+  polyprediction& operator=(const polyprediction&) = delete;
+  // ~polyprediction();
```
Comment (olgavrou) [inline vowpalwabbit/example.h:61]: nit: remove commented out destructor

**Review by olgavrou — APPROVED** [review]

**Review by jackgerrits — APPROVED** [review]

File: vowpalwabbit/nn.cc (hunk 1 above)
Comment (kumpera) [reply vowpalwabbit/nn.cc:398]: Not really, the livecycle of non-scalar predictions remains a mystery in this case.

File: vowpalwabbit/example.h (hunk 3 above)
Comment (kumpera) [reply vowpalwabbit/example.h]: done

**Pull request merged**